### PR TITLE
optimization for thousands of rules added one by one

### PR DESCRIPTION
**Pull request by GuillaumeGen**
Is this the same as #204 for you @francoisthire ?

### CODE REVIEW

File: kernel/signature.ml [hunk 1]
```diff
@@ -159,15 +159,15 @@ and add_rule_infos sg (lst:rule_infos list) : unit =
     in
     if infos.stat = Static && !fail_on_symbol_not_found
     then raise (Signature_error (CannotAddRewriteRules (r.l,r.cst)));
-    HId.replace env (id r.cst) {infos with rules = infos.rules @ rs; decision_tree= None}
+    HId.replace env (id r.cst) {infos with rules = rs @ infos.rules; decision_tree= None}
```
Comment (Gaspi) [inline kernel/signature.ml]: I think you need to `rev_append` here since rules are meant to be in reverse order. Besides it's faster.

Comment (francoisthire) [reply kernel/signature.ml]: You are right, I'll fix that.

Comment (francoisthire) [reply kernel/signature.ml]: You are right.

**Comment by francoisthire** [comment]
Well, I think this solves the one by one issue I mentioned but because you have a concatenation you are still in O(n^2) for other cases.

**Comment by GuillaumeGen** [comment]
Yeah, but I avoid to do flatten (which is basically a sequence of @), which can be overcostly, if you have to construct the tree several times.
Plus I avoid the `flatten(rev(rules))` lines, which are very tricky since it does not commute. I think they are in the wrong order in #204.

**Comment by GuillaumeGen** [comment]
Regarding the `rev_append` in the file `signature.ml`, it is not required with this version of the code, since I did a `rev_map` in the parser.
I find it easier to follow if all the list of rules are Fifo piles, but if you find it better that the output of the parser are in the order the user wrote her file, I also understand this and I can change it.

**Comment by francoisthire** [comment]
> Yeah, but I avoid to do flatten (which is basically a sequence of @), which can be overcostly, if you have to construct the tree several times.

I don't understand your point here. Why is it overcostly? Are you talking about a constant factor or a complexity problem?

Generation of trees is lazy. Once you need to generate a tree, you need to pay for the concatenation of rules anyway, you need to have a O(n) pass. And if you add m rules, you need again to do an O(n+m) pass.

> Plus I avoid the flatten(rev(rules)) lines, which are very tricky since it does not commute. I think they are in the wrong order in #204.

I don't see why. Can you give an example?

> Regarding the rev_append in the file signature.ml, it is not required with this version of the code, since I did a rev_map in the parser.

Yes, you are right. But this is terrible! There is now an implicit invariant which depends on how the parser output rules.. Moreover, it will break the printing since you did not modify the printing functions, hence rules will be printed in the wrong order if I am not mistaken.

A better way would be to do the rev only when the rules are given to Signature but I am not sure at all that it is a good way.

**Comment by GuillaumeGen** [comment]
I agree it is a linear factor, not a jump in complexity.

> > Plus I avoid the flatten(rev(rules)) lines, which are very tricky since it does not commute. I think they are in the wrong order in #204.
> 
> I don't see why. Can you give an example?

If you have for rules, added this way: `a b. c d.`, then, your signature will contain [[d,c],[b,a]], if you rev, then flatten, you get [b,a,d,c], which is not the order the rules were declared by the user.

> Yes, you are right. But this is terrible! There is now an implicit invariant which depends on how the parser output rules.. Moreover, it will break the printing since you did not modify the printing functions, hence rules will be printed in the wrong order if I am not mistaken.
Signature is already assuming that the parser is giving it the rule in the order it read it. I think that it does not change the quantity of order-preservation we are assuming.
Regarding printing, you are right.

**Comment by francoisthire** [comment]
> Signature is already assuming that the parser is giving it the rule in the order it read it. I think that it does not change the quantity of order-preservation we are assuming.
Regarding printing, you are right.

Ok, but maintaning the rules in the reverse order only for Signature is what I was qualifying as `terrible`. This reverse thing should be only inside the `Signature` module and not outside except if we realize that it is better for other applications but in that case, it should be explicit in the `Entry` type.

**Comment by GuillaumeGen** [comment]
I don't understand what you mean by "only for Signature". It is a lifo pile in all the code, if it already one at parsing time.

**Comment by francoisthire** [comment]
I think we can merge and close #204.

**Comment by rlepigre** [comment]
> Well, I think this solves the one by one issue I mentioned but because you have a concatenation you are still in O(n^2) for other cases. I did not test yet your implementation though.
> 
> You need to remove the `append` function to avoid any issue in the general case.

If you really need performance here you should implement a list data structure with constant time concatenation. I guess all you need is to iterate over the data structure.
```ocaml
type 'a app_list =
  | Nil
  | Cons   of 'a * 'a app_list
  | Append of 'a app_list * 'a app_list

let iter : ('a -> unit) -> 'a app_list -> unit = fun f l ->
  let rec iter ls =
    match ls with
    | []               -> ()
    | Nil        :: ls -> iter ls
    | Cns(x,l)   :: ls -> f x; iter (l :: ls)
    | App(l1,l2) :: ls -> iter (l1 :: l2 :: ls)
  in
  iter [l]
```

**Pull request merged**